### DISCUSSION TITLE
DMA problems with STM32F750

### DISCUSSION THREAD

**Discussion by HanSoke** [zephyrproject-rtos/zephyr#37890] [Q&A] [answered]
Hi,

I'm working with STM32F7508-DK Discovery kit and tried to port Zephyr for it. I have an application code which makes simple memory to memory DMA transfer and it seems to work occasionally. After transfer DMA driver triggers call back function without errors but nothing is copied to target buffer. If transfer is repeated in loop it might succeed sometimes.

Same code works perfectly with all other boards I have tested (nucleo-f411re, nucleo-f429zi, nucleo-l452re-p).

For devicetree and other config files I have used STM32F756 as a base. Only difference I noticed was size of the flash.

Does anyone have ideas how to proceed with this?

**Comment by FRASTM** [comment, marked as answer]
Yes, maybe you are facing a cache un-coherency on the stm32f7x.
Do you try to disable the DCache ?
Just for testing if it fix your pb, can you remove the SCB_EnableDCache in the **_soc/arm/st_stm32/stm32f7/soc.c_**:
```
	/*if (!(SCB->CCR & SCB_CCR_DC_Msk)) {
		SCB_EnableDCache();
	}*/
```

**Reply by HanSoke** [reply]
Thanks, it seems to be cache related. Removing the `SCB_EnableDCache` really fixes the problem.

**Reply by FRASTM** [reply]
One possible solution is to allocate buffers directly in the NOCACHE_MEMORY 
 see ./tests/drivers/dma/loop_transfer/src/test_dma_loop.c as an example
However for USERSPACE application, this is a bit more complicated, memcopy from DMA non-cached buffers to the user application ones (or extend the NON-Cached area like in in PR #37917
Another solution is to control the cache coherency when DMA is on, this is explained in the [AN4839: Level 1 cache on STM32F7 Series and STM32H7 Series](https://www.st.com/content/ccc/resource/technical/document/application_note/group0/08/dd/25/9c/4d/83/43/12/DM00272913/files/DM00272913.pdf/jcr:content/translations/en.DM00272913.pdf)
